schematic/store/synapse_tracker.py

```diff
@@ -0,0 +1,140 @@
+            retrieve_if_not_present: If True, retrieve the entity if it is not present in the cache.
+            download_location: The location to download the file to.
+            if_collision: The action to take if there is a collision when downloading the file.
+
+        Returns:
+            The Synapse entity.
+        """
+        entity = self.synapse_entities.get(synapse_id, None)
+
+        if entity is None or (download_file and not entity.path):
+            if not retrieve_if_not_present:
+                return None
+            entity = syn.get(
+                synapse_id,
+                downloadFile=download_file,
+                downloadLocation=download_location,
+                ifcollision=if_collision,
+            )
+        self.synapse_entities.update({synapse_id: entity})
+        return entity
+
+    def add(
+        self, synapse_id: str, entity: Union[Entity, Project, File, Folder, Schema]
+    ) -> None:
+        """Adds a Synapse entity to the cache.
+
+        Args:
+            synapse_id: The Synapse ID of the entity to add.
+            entity: The Synapse entity to add.
+        """
+        self.synapse_entities.update({synapse_id: entity})
+
+    def remove(self, synapse_id: str) -> None:
+        """Removes a Synapse entity from the cache.
+
+        Args:
+            synapse_id: The Synapse ID of the entity to remove.
+        """
+        self.synapse_entities.pop(synapse_id, None)
+
+    def search_local_by_parent_and_name(
+        self, name: str, parent_id: str
+    ) -> Union[Entity, Project, File, Folder, Schema, None]:
+        """
+        Searches the local cache for an entity with the given name and parent_id. The
+        is useful in situations where we might not have the ID of the resource, but we
+        do have the name and parent ID.
+
+        Args:
+            name: The name of the entity to search for.
+            parent_id: The parent ID of the entity to search for.
+
+        Returns:
+            The entity if it exists, otherwise None.
+        """
+        for entity in self.synapse_entities.values():
+            if entity.name == name and entity.parentId == parent_id:
+                return entity
+        return None
+
+    def get_project_headers(
+        self, syn: synapseclient.Synapse, current_user_id: str
+    ) -> List[Dict[str, str]]:
+        """Gets the paginated results of the REST call to Synapse to check what projects the current user has access to.
+
+        Args:
+            syn: A Synapse object
+            current_user_id: profile id for the user whose projects we want to get.
+
+        Returns:
+            A list of dictionaries matching <https://rest-docs.synapse.org/rest/org/sagebionetworks/repo/model/ProjectHeader.html>
+        """
+        project_headers = self.project_headers.get(current_user_id, None)
+        if project_headers:
+            return project_headers
+
+        all_results = syn.restGET(
```

Comment:
Nit: There is a syn._GET_Paginated if that makes it easier

Reply:
That, nor `_POST_paginated` would appear to work in this case as `nextPageToken` is a query parameter, instead of a piece of the body.

The return type of:
https://rest-docs.synapse.org/rest/GET/projects/user/principalId.html

Is different in it's format to the `_GET_Paginated` expected type of:
https://rest-docs.synapse.org/rest/GET/teamMembers/id.html